File: library/src/main/java/com/luck/picture/lib/PictureSelector.java
package com.luck.picture.lib;

import android.app.Activity;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import java.lang.ref.WeakReference;

/**
 * @author：luck
 * @date：2017-5-24 22:30
 * @describe：PictureSelector
 */

public final class PictureSelector {

    private final WeakReference<Activity> mActivity;
    private final WeakReference<Fragment> mFragment;

    private PictureSelector(Activity activity) {
        this(activity, null);
    }

    private PictureSelector(Fragment fragment) {
        this(fragment.getActivity(), fragment);
    }

    private PictureSelector(Activity activity, Fragment fragment) {
        mActivity = new WeakReference<>(activity);
        mFragment = new WeakReference<>(fragment);
    }

    /**
     * Start PictureSelector for Activity.
     *
     * @param activity
     * @return PictureSelector instance.
     */
    public static PictureSelector create(Activity activity) {
        return new PictureSelector(activity);
    }

    /**
     * Start PictureSelector for Fragment.
     *
     * @param fragment
     * @return PictureSelector instance.
     */
    public static PictureSelector create(Fragment fragment) {
        return new PictureSelector(fragment);
    }

    /**
     * @param chooseMode Select the type of picture you want，all or Picture or Video .
     * @return LocalMedia PictureSelectionModel
     * Use {@link com.luck.picture.lib.config.SelectMimeType}
     */
    public PictureSelectionModel openGallery(int chooseMode) {
        return new PictureSelectionModel(this, chooseMode);
    }

    /**
     * @param chooseMode Select the type of picture you want，Picture or Video.
     * @return LocalMedia PictureSelectionModel
     * Use {@link com.luck.picture.lib.config.SelectMimeType}
     */
    public PictureSelectionModel openCamera(int chooseMode) {
        return new PictureSelectionModel(this, chooseMode, true);
    }

    /**
     * @return Activity.
     */
    @Nullable
    Activity getActivity() {
        return mActivity.get();
    }

    /**
     * @return Fragment.
     */
    @Nullable
    Fragment getFragment() {
        return mFragment != null ? mFragment.get() : null;
    }

}

File: selector/src/main/java/com/luck/picture/lib/model/PictureSelector.kt
package com.luck.picture.lib.model

import android.app.Activity
import android.content.Context
import android.content.Intent
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import com.luck.picture.lib.config.MediaType
import com.luck.picture.lib.constant.SelectorConstant
import com.luck.picture.lib.entity.LocalMedia
import java.lang.ref.SoftReference

/**
 * @author：luck
 * @date：2017-5-24 22:30
 * @describe：PictureSelector
 */

class PictureSelector {

    companion object {
        fun create(context: Context): PictureSelector {
            return PictureSelector(context as Activity)
        }

        fun create(activity: FragmentActivity): PictureSelector {
            return PictureSelector(activity)
        }

        fun create(fragment: Fragment): PictureSelector {
            return PictureSelector(fragment)
        }

        fun obtainSelectResults(intent: Intent?): ArrayList<LocalMedia> {
            intent?.let {
                return intent.getParcelableArrayListExtra<LocalMedia>(SelectorConstant.KEY_EXTRA_RESULT) as ArrayList<LocalMedia>
            }
            return arrayListOf()
        }
    }

    /**
     * open gallery source
     * use [MediaType]
     */
    fun openGallery(mediaType: MediaType): SelectionMainModel {
        return SelectionMainModel(this, mediaType)
    }

    /**
     * From system album Select the type of images you want，all or images or video or audio
     * use [MediaType]
     */
    fun openSystemGallery(mediaType: MediaType): SelectionSystemModel {
        return SelectionSystemModel(this, mediaType)
    }

    /**
     * only use camera，images or video or audio
     * use [MediaType]
     */
    fun openCamera(mediaType: MediaType): SelectionCameraModel {
        return SelectionCameraModel(this, mediaType)
    }

    /**
     * query the type of images you want，all or images or video or audio
     * use [MediaType]
     */
    fun dataSource(mediaType: MediaType): SelectionQueryModel {
        return SelectionQueryModel(this, mediaType)
    }

    /**
     * preview mode
     */
    fun openPreview(): SelectionPreviewModel {
        return SelectionPreviewModel(this)
    }

    private var mActivity: SoftReference<Activity>? = null
    private var mFragment: SoftReference<Fragment>? = null

    constructor(activity: Activity) : this(activity, null)

    constructor(fragment: Fragment) : this(fragment.activity, fragment)

    constructor(activity: Activity?, fragment: Fragment?) {
        mActivity = SoftReference(activity)
        mFragment = SoftReference(fragment)
    }

    internal fun getActivity(): Activity? {
        return mActivity?.get()
    }

    internal fun getFragment(): Fragment? {
        return if (mFragment != null) mFragment?.get() else null
    }
}